Sources/ProtobufferCoding/Decoder/KeyedProtoDecodingContainer.swift

```diff
@@ -0,0 +1,328 @@
+        let keyValue = try extractIntValue(from: key)
+        // as of Proto3, repeated values of scalar numeric types are always encoded as packed
+        // thus, there will only be one value for the given key,
+        // containing n number
+        if let value = data[keyValue]?.last {
+            return try decodeRepeatedInt32(value)
+        }
+        throw ProtoError.decodingError("No data for given key")
+    }
+
+    func decode(_ type: [Int64].Type, forKey key: Key) throws -> [Int64] {
+        let keyValue = try extractIntValue(from: key)
+        // as of Proto3, repeated values of scalar numeric types are always encoded as packed
+        // thus, there will only be one value for the given key,
+        // containing n number
+        if let value = data[keyValue]?.last {
+            return try decodeRepeatedInt64(value)
+        }
+        throw ProtoError.decodingError("No data for given key")
+    }
+
+    func decode(_ type: [UInt32].Type, forKey key: Key) throws -> [UInt32] {
+        let keyValue = try extractIntValue(from: key)
+        // as of Proto3, repeated values of scalar numeric types are always encoded as packed
+        // thus, there will only be one value for the given key,
+        // containing n number
+        if let value = data[keyValue]?.last {
+            return try decodeRepeatedUInt32(value)
+        }
+        throw ProtoError.decodingError("No data for given key")
+    }
+
+    func decode(_ type: [UInt64].Type, forKey key: Key) throws -> [UInt64] {
+        let keyValue = try extractIntValue(from: key)
+        // as of Proto3, repeated values of scalar numeric types are always encoded as packed
+        // thus, there will only be one value for the given key,
+        // containing n number
+        if let value = data[keyValue]?.last {
+            return try decodeRepeatedUInt64(value)
+        }
+        throw ProtoError.decodingError("No data for given key")
+    }
+
+    func decode(_ type: [Data].Type, forKey key: Key) throws -> [Data] {
+        let keyValue = try extractIntValue(from: key)
+        if let values = data[keyValue] {
+            // the data is already [Data] :D
+            return values
+        }
+        throw ProtoError.decodingError("No data for given key")
+    }
+
+    func decode(_ type: [String].Type, forKey key: Key) throws -> [String] {
+        let keyValue = try extractIntValue(from: key)
+        if let values = data[keyValue] {
+            return decodeRepeatedString(values)
+        }
+        throw ProtoError.decodingError("No data for given key")
+    }
+
+    // swiftlint:disable cyclomatic_complexity
+    func decode<T>(_ type: T.Type, forKey key: Key) throws -> T where T: Decodable {
+        // we need to switch here to also be able to decode structs with generic types
+        // if struct has generic type, this will always end up here
+        if T.self == Data.self, let value = try decode(Data.self, forKey: key) as? T {
```

Comment:
Thank you so much for the explanatory comment

Reply:
BTW this seems like a place for using gyb: https://nshipster.com/swift-gyb/

Reply:
Yes, indeed. Never used it before, so I'll take a closer look at it.
Thanks for the comment 🙌